frontend/test/pytest/test_jax_dynamic_api.py

```diff
@@ -505,7 +530,7 @@ def f(sz: int):
         input_a = jnp.ones([sz + 1], dtype=float)
         input_b = jnp.ones([sz + 2], dtype=float)
 
-        @for_loop(0, 10, 1, experimental_preserve_dimensions=True)
+        @for_loop(0, 10, 1, experimental_preserve_dimensions=False)
```

Comment:
Is `experimental_preserve_dimensions=False` not default?

Reply:
No, it defaults to True.

Reply:
Is it not confusing for users to have an experimental feature turned on by default?

Reply:
I think maybe we should change this name. It is not an experimental feature. Here are the semantics:

```
This PR sets the following semantic of loops, depending on the value of the already-existing experimental_preserve_dimensions flag:
True (the default): all dynamic dimension variables will be handled as Jaxpr constants. So
    Mixing argument and captured arrays is possible.
    No dimension modification is allowed inside loops.
False would handle all the dimensions as Jaxpr implicit arguments.
    Mixing argument and captured arrays is not possible.
    Dimension modification within the loops are allowed.
```

Pinging @josh146 here. Maybe a better name would be to just drop the `experimental` part of it?

Reply:
Yes this is a great point. Maybe something like `experimental_allow_shape_changes`?

Also, I remember that originally this flag also determined whether two arrays with the same dynamic shape could be operated on together. I don't see it listed in the semantics, does that restriction still apply?

Reply:
> Also, I remember that originally this flag also determined whether two arrays with the same dynamic shape could be operated on together. I don't see it listed in the semantics, does that restriction still apply?

You mean having an operation with two operands with dynamic shape?

Reply:
> You mean having an operation with two operands with dynamic shape?

Exactly, two operands of the same dynamic shape. In the original PR that is the restriction the flag was about, before Sergei also discovered the issue about capturing outside values.

Reply:
Talked to Josh and he will update the naming/docs in a follow up :) We will need clarity about that second point though wether the semantics of the flag also influence wether two arrays can be used in a binary operator.

For now we can merge this PR 👍

Reply:
See https://github.com/PennyLaneAI/catalyst/pull/904 :) Feel free to directly push to this PR.